### PR TITLE
Fix log_loss and add test for multiclass

### PR DESCRIPTION
**Pull request by glennq**
In `MLPClassifier`, loss is set to `'log_loss'` for both multiclass case and multilabel case. 

The `log_loss` implementation in `neural_network/base.py` is in fact binary cross entropy, but for multiclass case where `self.out_activation_` is set to `'softmax'`, it does not make sense to use binary cross entropy loss. I think categorical cross entropy, or otherwise called negative log likelihood should be used instead.

### CODE REVIEW

File: sklearn/neural_network/base.py [hunk 1]
```diff
@@ -189,8 +189,40 @@ def log_loss(y_true, y_prob):
...
+    if y_prob.shape[1] == 1:
+        y_prob = np.append(1 - y_prob, y_prob, axis=1)
+
+    if y_true.shape[1] == 1:
+        y_true = np.append(1 - y_true, y_true, axis=1)
+
+    return -np.sum(y_true * np.log(y_prob)) / y_prob.shape[0]
+
+
+def binary_log_loss(y_true, y_prob):
```
Comment (amueller) [inline sklearn/neural_network/base.py]: Don't you think calling it multilabel_log_loss would be more informative?

**Comment by amueller** [comment]
I agree, that's a bug.
It would be good to have a non-regression test.

**Comment by amueller** [comment]
I'm not sure if it is better to work on this branch or on #3204. But we can fix this here for now.

File: sklearn/neural_network/base.py (hunk 1 above)
Comment (glennq) [inline sklearn/neural_network/base.py]: You mean renaming `binary_log_loss` to `multilabel_log_loss`? This makes sense, it's just that this loss function is actually used in both binary and multilabel case since I made minimal change to `BaseMultilayerPerceptron._initialize()`. I took the 'binary' here from 'binary cross-entropy'.

**Comment by glennq** [comment]
I'm also confused about which pull request is going to be merged to master, but I guess we should first merge the two pull requests.

**Comment by amueller** [comment]
Unfortunately that is not entirely decided. #3204 is more "up to date" but I'm not sure if all the additions there are useful.
It would be good to benchmark / review the implemented learning algorithms before deciding which way to go.

**Comment by glennq** [comment]
I see. I was trying to compare the two pull requests through github but couldn't get an informative comparison result. 
I guess the main difference is the newly added `learning_rate_class` which implements a few learning rate adaption schemes for SGD.
So I think you were talking about benchmark on these algorithm. This makes sense.

I'm just wondering if it would be better to merge an extensible minimal working implementation first, and consider new features later, since this pull request has been around for nearly two years...

**Comment by amueller** [comment]
I mostly agree with your assessment. The problem is that we need working default values, and if the defaults just not work, people might be discouraged from using it.
So we need at least one well-working example. And we probably need to include early stopping.

I just wrote an email to some people who have worked on this and on deep learning in general to ask their opinion.

Apart from benchmarking the learning schedules (these are not really learning rates imho), the other thing would be investigating early stopping.
I think it is not included in my branch yet, but there should be an option to split the provided dataset into a test and validation set, and do early stopping on the validation set.
It would be great if you could provide an example where this is beneficial.
